DylibBinder/DBTypeDeclaration.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using DylibBinder;
+using SwiftReflector;
+using SwiftReflector.Inventory;
+
+namespace DylibBinder {
+	public class DBTypeDeclaration {
+		public DBTypeDeclaration (ProtocolContents protoContents)
+		{
+			Kind = "protocol";
+			Name = protoContents.Name.ToString ();
+			Funcs = new DBFuncs (protoContents);
+			ApplyInouts ();
+		}
+
+		public DBTypeDeclaration (ClassContents classContents)
+		{
+			if (classContents.Name.IsClass)
+				Kind = "class";
+			else if (classContents.Name.IsEnum)
+				Kind = "enum";
+			else
+				Kind = "struct";
+
+			Name = classContents.Name.ToFullyQualifiedName ();
+			Funcs = new DBFuncs (classContents);
+			Properties = new DBProperties (classContents);
+			GenericParameters = new DBGenericParameters (this);
+			InnerTypes = new DBInnerTypes (classContents.Name);
+			AssociatedTypes = new DBAssociatedTypes ();
+			Elements = new DBElements ();
+			ApplyInouts ();
+		}
+
+		public string Kind { get; }
+		public string Name { get; }
+		public string Accessibility { get; } = "Public";
+		public string IsObjC { get; } = "False";
+		public string IsFinal { get; } = "False";
+		public string IsDeprecated { get; } = "False";
+		public string IsUnavailable { get; } = "False";
+
+		public DBFuncs Funcs { get; }
+		public DBProperties Properties { get; }
+		public DBGenericParameters GenericParameters { get; }
+		public DBAssociatedTypes AssociatedTypes { get; }
+		public DBInnerTypes InnerTypes { get; }
+		public DBElements Elements { get; }
+
+		void ApplyInouts ()
+		{
+			foreach (var func in Funcs.Funcs) {
+				ApplyInoutsToParameterLists (func.ParameterLists);
+			}
+			// protocols currently only hold funcs so they should finish here
+			if (Kind == "protocol")
+				return;
+
+			foreach (var property in Properties.Properties) {
+				ApplyInoutsToParameterLists (property.Getter.ParameterLists);
+				if (property.Setter != null) {
+					ApplyInoutsToParameterLists (property.Setter.ParameterLists);
+				}
+			}
+		}
+
+		void ApplyInoutsToParameterLists (DBParameterLists parameterLists)
+		{
+			foreach (var parameterList in parameterLists.ParameterLists) {
+				ApplyInoutsToParameterList (parameterList);
+			}
+		}
+
+		void ApplyInoutsToParameterList (DBParameterList parameterList)
+		{
+			foreach (var parameter in parameterList.Parameters) {
+				if (parameter.HasInstance || parameter.IsConstructor) {
+					ApplyInoutsToParameter (parameter);
+				}
+			}
+		}
+
+		void ApplyInoutsToParameter (DBParameter parameter)
+		{
+			var sb = new StringBuilder ();
+			sb.Append ($"inout {Name}{GenericParametersToString ()}");
+			parameter.Type = parameter.Type.Insert (0, sb.ToString ());
```

Comment:
This looks 🐟  care to explain?

Reply:
Started writing why I thought I needed to do it this way, but I think through writing it down, I found a more clear way to do this part!

Reply:
I take this back now. I went to try to implement that idea I had but it did not work. If you have a better suggestion for this then let me know! @mandel-macaque @stephen-hawley 

The purpose of this area is to add the string 
```inout {name of the nominalType}{<NominalType's GenericParameters if any>}```
 to the parameters of methods and properties that are instances or are constructors. (constructors need an additional ".Type" at the end of this string)

When I am getting the method/property's information, I know if it should have this extra parameter with this special string, but I do not know the nominal type's generic parameters yet. That is why I am getting the generic parameters first and then going back to the parameters and adding the special string above.

### Process (the relevant parts for clarification)

- Get TypeDeclaration information
    - Get Methods inside this TypeDeclaration
         -  Get each Method
             - Get Parameters for this Method
             - Get Generic Parameters used in this Method
    - Get Properties inside this TypeDeclaration
        - Get each property
            - Get the type of this Property
            - Get Generic Parameters used in this Property 
    - Look at the Generic Parameters in this Typedeclaration's Methods and Properties and get the top level ones
    - **Go back through all the Methods' parameters and return types and Properties' types and add special string including the generic parameters to the inouts**